### PR TITLE
feat(api): OnDecision Notification Listener

### PR DESCRIPTION
**Pull request by mfahadahmed**
## Summary
- Implemented onDecisionListener in following APIs:
isfeatureEnabled
getEnabledFeatures
getFeatureVariableString
getFeatureVariableBoolean
getFeatureVariableInteger
getFeatureVariableDouble 

## Test plan
- Unit testss for OnDecisionListener

### CODE REVIEW

File: packages/optimizely-sdk/lib/utils/enums/index.js [hunk 1]
```diff
@@ -177,6 +177,13 @@ exports.NODE_CLIENT_VERSION = '3.0.1';
 exports.NOTIFICATION_TYPES = {
   ACTIVATE: 'ACTIVATE:experiment, user_id,attributes, variation, event',
   TRACK: 'TRACK:event_key, user_id, attributes, event_tags, event',
+  ON_DECISION: 'ON_DECISION:type, user_id, attributes, decision_info',
+};
+
+exports.ON_DECISION_NOTIFICATION_TYPES = {
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/utils/enums/index.js]: Lets call this `DECISION_INFO_TYPES`

They aren't really `NOTIFICATION_TYPES` like the ones above.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 2]
```diff
@@ -304,7 +305,21 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
         return null;
       }
 
-      return this.decisionService.getVariation(experimentKey, userId, attributes);
+      var variationKey = this.decisionService.getVariation(experimentKey, userId, attributes);
+      this.notificationCenter.sendNotifications(
+        enums.NOTIFICATION_TYPES.ON_DECISION,
+        {
+          type: ON_DECISION_NOTIFICATION_TYPES.EXPERIMENT_VARIATION,
+          user_id: userId,
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.js]: lets use `camelCase` property names to be consistent with other notification events.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 3]
```diff
@@ -463,20 +478,45 @@ Optimizely.prototype.isFeatureEnabled = function (featureKey, userId, attributes
...
-    this.logger.log(LOG_LEVEL.INFO, sprintf(LOG_MESSAGES.FEATURE_NOT_ENABLED_FOR_USER, MODULE_NAME, featureKey, userId));
-    return false;
+    
+    if (!featureEnabled) {
+      this.logger.log(LOG_LEVEL.INFO, sprintf(LOG_MESSAGES.FEATURE_NOT_ENABLED_FOR_USER, MODULE_NAME, featureKey, userId));
+      featureEnabled = false;
+    }
+
+    var decisionSource = decision.decisionSource || DECISION_SOURCES.ROLLOUT;
+    decisionSource += decisionSource === DECISION_SOURCES.EXPERIMENT ? sprintf(' {%s}', decision.experiment.key) : '';
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Lets keep this the same as `decision.decisionSource` and add a new field to `decision_info` 
```
      {
        type: ON_DECISION_NOTIFICATION_TYPES.FEATURE,
        user_id: userId,
        attributes: attributes,
        decisionInfo: {
          featureKey: featureKey,
          featureEnabled: featureEnabled,
          source: 'ROLLOUT',
          experimentKey: null,
        }
```

and

```
      {
        type: ON_DECISION_NOTIFICATION_TYPES.FEATURE,
        user_id: userId,
        attributes: attributes,
        decisionInfo: {
          featureKey: featureKey,
          featureEnabled: featureEnabled,
          source: 'EXPERIMENT',
          experimentKey: 'feature_exp_key',
        }
```

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 4]
```diff
@@ -463,20 +478,45 @@ Optimizely.prototype.isFeatureEnabled = function (featureKey, userId, attributes
...
+
+    console.log(sprintf('Feature: %s - Enabled: %s, Source: %s', featureKey, featureEnabled, decisionSource))
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.js]: remove this console log.

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 5]
```diff
@@ -565,15 +605,37 @@ Optimizely.prototype._getFeatureVariableForType = function(featureKey, variableK
...
+    featureEnabled = decision.variation.featureEnabled;
     this.logger.log(LOG_LEVEL.INFO, sprintf(LOG_MESSAGES.USER_RECEIVED_VARIABLE_VALUE, MODULE_NAME, variableKey, featureFlag.key, variableValue, userId));
   } else {
     variableValue = variable.defaultValue;
     this.logger.log(LOG_LEVEL.INFO, sprintf(LOG_MESSAGES.USER_RECEIVED_DEFAULT_VARIABLE_VALUE, MODULE_NAME, userId, variableKey, featureFlag.key));
   }
 
-  return projectConfig.getTypeCastValue(variableValue, variableType, this.logger);
+  var decisionSource = decision.decisionSource || DECISION_SOURCES.ROLLOUT;
+  decisionSource += decisionSource === DECISION_SOURCES.EXPERIMENT ? sprintf(' {%s}', decision.experiment.key) : '';
```
Comment (jordangarcia) [inline packages/optimizely-sdk/lib/optimizely/index.js]: Same thing as above.  Let's keep `decisionSource` unmodified and add `experimentKey` to `decisionInfo`

**Comment by aliabbasrizvi** [comment]
@mfahadahmed I am going to request you to break this change into 2 separate PRs at least. One for feature enabled and one for feature variable. It will make reviewing go faster.

File: packages/optimizely-sdk/lib/utils/enums/index.js [hunk 6]
```diff
@@ -177,6 +178,13 @@ exports.NODE_CLIENT_VERSION = '3.0.1';
 exports.NOTIFICATION_TYPES = {
   ACTIVATE: 'ACTIVATE:experiment, user_id,attributes, variation, event',
   TRACK: 'TRACK:event_key, user_id, attributes, event_tags, event',
+  ON_DECISION: 'ON_DECISION:type, user_id, attributes, decision_info',
+};
+
+exports.DECISION_INFO_TYPES = {
+  EXPERIMENT_VARIATION: 'experiment_variation',
```
Comment (msohailhussain) [inline packages/optimizely-sdk/lib/utils/enums/index.js]: Why it's experiment_variation? It should be experiment.

File: packages/optimizely-sdk/lib/utils/enums/index.js [hunk 7]
```diff
@@ -177,6 +178,13 @@ exports.NODE_CLIENT_VERSION = '3.0.1';
 exports.NOTIFICATION_TYPES = {
   ACTIVATE: 'ACTIVATE:experiment, user_id,attributes, variation, event',
   TRACK: 'TRACK:event_key, user_id, attributes, event_tags, event',
+  ON_DECISION: 'ON_DECISION:type, user_id, attributes, decision_info',
```
Comment (msohailhussain) [inline packages/optimizely-sdk/lib/utils/enums/index.js]: It's been rename to Decision?

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 8]
```diff
@@ -304,7 +305,21 @@ Optimizely.prototype.getVariation = function(experimentKey, userId, attributes)
         return null;
       }
 
-      return this.decisionService.getVariation(experimentKey, userId, attributes);
+      var variationKey = this.decisionService.getVariation(experimentKey, userId, attributes);
+      this.notificationCenter.sendNotifications(
+        enums.NOTIFICATION_TYPES.ON_DECISION,
+        {
+          type: DECISION_INFO_TYPES.EXPERIMENT_VARIATION,
```
Comment (msohailhussain) [inline packages/optimizely-sdk/lib/optimizely/index.js]: There is no `type` EXPERIMENT_VARIATION

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 2 above)
Comment (msohailhussain) [reply packages/optimizely-sdk/lib/optimizely/index.js]: If we are using camelCase, then it must be reflected here as well for `Activate`, `Track` and `ON_DECISION/Decision`
https://github.com/optimizely/javascript-sdk/pull/240/files#diff-79c0858a5dc2611f6570afc1265258daR179

File: packages/optimizely-sdk/lib/optimizely/index.js [hunk 9]
```diff
@@ -34,6 +34,7 @@ var LOG_MESSAGES = enums.LOG_MESSAGES;
 var MODULE_NAME = 'OPTIMIZELY';
 var DECISION_SOURCES = enums.DECISION_SOURCES;
 var FEATURE_VARIABLE_TYPES = enums.FEATURE_VARIABLE_TYPES;
+var DECISION_INFO_TYPES = enums.DECISION_INFO_TYPES;
```
[the remark this answers is not in the document]
Comment (jordangarcia) [reply packages/optimizely-sdk/lib/optimizely/index.js]: I think `DECISION_INFO_TYPES` is fine.  It was confusing before when it was `DECISION_NOTIFICATION_TYPES`

File: packages/optimizely-sdk/lib/optimizely/index.js (hunk 8 above)
Comment (jordangarcia) [reply packages/optimizely-sdk/lib/optimizely/index.js]: I believe @aliabbasrizvi changed this to be just `EXPERIMENT` in the design doc

File: packages/optimizely-sdk/lib/utils/enums/index.js (hunk 7 above)
Comment (jordangarcia) [reply packages/optimizely-sdk/lib/utils/enums/index.js]: I dont think this change has been finalized. I'd keep it how it is today for consistency with the other `NOTIFICATION_TYPES`

File: packages/optimizely-sdk/lib/utils/enums/index.js (hunk 6 above)
Comment (jordangarcia) [reply packages/optimizely-sdk/lib/utils/enums/index.js]: Would you mind checking with @aliabbasrizvi on whether this should be `experiment_variation` or `experiment` . 

Right now it's still `experiment_variation` in the design doc

Comment (msohailhussain) [reply packages/optimizely-sdk/lib/utils/enums/index.js]: In th example, `experiment` was used. Added comment in the doc.

**Comment by mfahadahmed** [comment]
@aliabbasrizvi @jordangarcia @msohailhussain  Closing this PR as this change has been split into PR #244 , #245 and #246 now.

**Pull request closed without merging**